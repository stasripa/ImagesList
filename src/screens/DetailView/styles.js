import { Dimensions, StyleSheet } from 'react-native'
const { width, height } = Dimensions.get('window')

const styles: any = StyleSheet.create({
  container: {
    backgroundColor: '#000',
    flex: 1,
  },
  imageContainer: {
    flex: 3,
    flexDirection: 'column',
    justifyContent: 'center',
    alignItems: 'center',
  },
  imageStyle: {
    flex: 1,
    width: width,
    height: height,
  },
  smallImage: {
    width: 150,
    height: 150,
  },
  backButton: {
    position: 'absolute',
    left: 5,
    top: 5,
  },
  spinner: {
    position: 'absolute',
  },
  detailView: {
    backgroundColor: 'rgba(0, 0, 0, 0.2)',
    padding: 8,
    position: 'absolute',
    bottom: 0,
    width: '100%',
    left: 0,
    flexDirection: 'row',
    justifyContent: 'space-between'
  },
  detailViewImage: {
    width: 50,
    height: 50,
  },
  text: {
    color: 'white',
  },
  row: {
    flexDirection: 'row',
  },
  filterItem: {
    marginHorizontal: 8,
  },
  filtersContainer: {
    flex: 1,
    height: 50,
    alignContent: 'center',
    backgroundColor: 'rgba(0, 0, 0, 0.2)',
    marginBottom: 58,
  },
  activityIndicatorViewStyle: {
    position: 'absolute',
    width: '100%',
    height: '100%',
    alignItems: 'center',
    justifyContent: 'center',
    borderRadius: 20,
    backgroundColor: 'rgba(0, 0, 0, 0.2)',
  },
})
export default styles
