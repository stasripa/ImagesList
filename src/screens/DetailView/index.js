// @flow
import * as React from 'react'
import {
  ActivityIndicator,
  View,
  ScrollView,
  Image,
  Dimensions,
  TouchableOpacity,
} from 'react-native'

import styles from './styles'
import DetailsFooter from './components/DetailsFooter'
import ImageZoom from 'react-native-image-pan-zoom'
import {
  ColorMatrix,
  concatColorMatrices,
  grayscale,
  saturate,
  sepia,
  polaroid,
  normal,
  hueRotate,
  invert,
  cool,
  warm,
  technicolor,
  browni,
  vintage,
  night,
  predator,
  lsd,
} from 'react-native-color-matrix-image-filters'

const { width, height } = Dimensions.get('window')

type Props = {
  isLoading: boolean,
  shareCallback: Function,
  pictureDetails: Object,
}

// TODO: it would be great to see here loader, pinch to zoom here and pan

class DetailView extends React.PureComponent<Props> {

  state = {
    filter: [sepia()],
    showFilters: false,
  }

  toggleFilters = () => {
    this.setState({showFilters: !this.state.showFilters})
  }

  renderImage = (isSmall: boolean = false) => {
    const { pictureDetails } = this.props
    return (
      <Image
        source={{uri: pictureDetails.full_picture}}
        style={isSmall ? styles.smallImage : styles.imageStyle} />
    )
  }

  renderZoomedImage = () => {
    const { showFilters } = this.state
    if (showFilters) {
      return null
    }

    return (
      <ImageZoom
        cropWidth={width}
        cropHeight={height}
        imageWidth={width}
        imageHeight={height}
      >
        {this.renderImage()}
      </ImageZoom>
    )
  }

  renderAppliedFilter = () => {
    const { filter, showFilters } = this.state
    if (!showFilters) {
      return null
    }

    return (
      <ColorMatrix
        matrix={concatColorMatrices([filter])}>
        {this.renderImage()}
      </ColorMatrix>
    )
  }

  renderFilterItem = (filter) => {
    return (
      <View style={styles.filterItem}>
        <TouchableOpacity onPress={()=> this.applyFilter(filter)}>
          <ColorMatrix
            matrix={concatColorMatrices([filter()])}>
            {this.renderImage(true)}
          </ColorMatrix>
        </TouchableOpacity>
      </View>
    )
  }

  renderFiltersList = () => {
    const { showFilters } = this.state
    if (!showFilters) return null

      return (
        <ScrollView style={styles.filtersContainer}
                    horizontal>
          {this.renderFilterItem(sepia)}
          {this.renderFilterItem(saturate)}
          {this.renderFilterItem(grayscale)}

          {this.renderFilterItem(polaroid)}
          {this.renderFilterItem(normal)}
          {this.renderFilterItem(hueRotate)}
          {this.renderFilterItem(invert)}
          {this.renderFilterItem(cool)}
          {this.renderFilterItem(warm)}
          {this.renderFilterItem(technicolor)}
          {this.renderFilterItem(browni)}
          {this.renderFilterItem(vintage)}
          {this.renderFilterItem(night)}
          {this.renderFilterItem(predator)}
          {this.renderFilterItem(lsd)}
        </ScrollView>
      )
  }

  applyFilter = (filter) => {
    this.setState({filter: [filter()]})
  }

  render () {
    const { showFilters } = this.state
    const { isLoading, shareCallback, pictureDetails } = this.props
    return (
      <View style={styles.container}>
        <View style={styles.imageContainer}>
          {showFilters ? this.renderAppliedFilter() : this.renderZoomedImage()}
        </View>
        {this.renderFiltersList()}
        <DetailsFooter
          pictureDetails={pictureDetails}
          shareCallback={shareCallback}
          toggleFilters={this.toggleFilters}
        />
        { isLoading &&
          <View style={styles.activityIndicatorViewStyle}>
            <ActivityIndicator animating={true} size={80}/>
          </View>
        }
      </View>
    )
  }
}

export default DetailView
