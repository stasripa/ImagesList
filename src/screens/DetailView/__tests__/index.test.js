import React from 'react'
import DetailView from '../index'
// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer'

const colorSwitchCallback = jest.fn()
const shareCallback = jest.fn()
const imageUrl = 'https://drscdn.500px.org/photo/247377659/m%3D900_k%3D1_a%3D1/v2?client_application_id=27071&webp=true&sig=b5b4070666461be985e7890d9c3b26054ae52c8c882fe9b6120bb2c627b53204'

it('renders correctly', () => {
  const tree = renderer.create(<DetailView
    imageUrl={imageUrl}
    isLoading={false}
    imageId={247377659}
    colorSwitchCallback={colorSwitchCallback}
    shareCallback={shareCallback} />).toJSON()
  expect(tree).toMatchSnapshot()
})