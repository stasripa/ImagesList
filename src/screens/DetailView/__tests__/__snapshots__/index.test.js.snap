// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#000",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "flexDirection": "column",
        "justifyContent": "center",
      }
    }
  >
    <View
      contentAspectRatio={undefined}
      enableResistance={true}
      enableScale={true}
      enableTransform={false}
      enableTranslate={true}
      maxOverScrollDistance={20}
      maxScale={1}
      onLayout={[Function]}
      onSingleTapConfirmed={undefined}
      onTransformGestureReleased={undefined}
      onViewTransformed={undefined}
      style={
        Object {
          "flex": 1,
          "height": 675,
          "width": 675,
        }
      }
    >
      <View
        style={
          Object {
            "flex": 1,
            "transform": Array [
              Object {
                "scale": 1,
              },
              Object {
                "translateX": 0,
              },
              Object {
                "translateY": 0,
              },
            ],
          }
        }
      >
        <Image
          capInsets={
            Object {
              "bottom": 0.1,
              "left": 0.1,
              "right": 0.1,
              "top": 0.1,
            }
          }
          enableScale={true}
          enableTransform={true}
          enableTranslate={true}
          onLoad={[Function]}
          onLoadStart={[Function]}
          resizeMode="contain"
          source={
            Object {
              "uri": "https://drscdn.500px.org/photo/247377659/m%3D900_k%3D1_a%3D1/v2?client_application_id=27071&webp=true&sig=b5b4070666461be985e7890d9c3b26054ae52c8c882fe9b6120bb2c627b53204",
            }
          }
          style={
            Array [
              Object {
                "flex": 1,
                "height": 675,
                "width": 675,
              },
              Object {
                "backgroundColor": "transparent",
              },
            ]
          }
        />
      </View>
    </View>
    <ActivityIndicator
      animating={false}
      color="#FFF"
      hidesWhenStopped={true}
      size="large"
      style={
        Object {
          "position": "absolute",
        }
      }
    />
  </View>
</View>
`;
